homeassistant/components/automation/trace.py

```diff
@@ -134,35 +125,14 @@ def as_short_dict(self) -> dict[str, Any]:
         return result
 
 
-class LimitedSizeDict(OrderedDict):
-    """OrderedDict limited in size."""
-
-    def __init__(self, *args, **kwds):
-        """Initialize OrderedDict limited in size."""
-        self.size_limit = kwds.pop("size_limit", None)
-        OrderedDict.__init__(self, *args, **kwds)
-        self._check_size_limit()
-
-    def __setitem__(self, key, value):
-        """Set item and check dict size."""
-        OrderedDict.__setitem__(self, key, value)
-        self._check_size_limit()
-
-    def _check_size_limit(self):
-        """Check dict size and evict items in FIFO order if needed."""
-        if self.size_limit is not None:
-            while len(self) > self.size_limit:
-                self.popitem(last=False)
-
-
 @contextmanager
 def trace_automation(hass, unique_id, config, context):
     """Trace action execution of automation with automation_id."""
     automation_trace = AutomationTrace(unique_id, config, context)
     trace_id_set((unique_id, automation_trace.run_id))
 
     if unique_id:
-        automation_traces = hass.data[DATA_AUTOMATION_TRACE]
+        automation_traces = hass.data[DATA_TRACE]
         if unique_id not in automation_traces:
             automation_traces[unique_id] = LimitedSizeDict(size_limit=STORED_TRACES)
         automation_traces[unique_id][automation_trace.run_id] = automation_trace
```

Comment:
We should abstract away how traces are stored from `automation` and `script` integrations.

```suggestion
        async_store_trace(hass, DOMAIN, automation_trace)
```

`async_store_trace` would be a helper from `trace` integration. It can then pull the unique_id from the trace itself and store it in the right place.

Reply:
Fixed in #48276